storage/badger/execution_data_tracker.go

```diff
@@ -0,0 +1,446 @@
+// to block heights up to and including the given height,
+// and updates the latest pruned height value.
+// It locks the ExecutionDataTracker and ensures that no other writes
+// can occur during the pruning.
+// It is up to the caller to ensure that this is never
+// called with a value higher than the fulfilled height.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) PruneUpToHeight(height uint64) error {
+	blobRecordPrefix := []byte{storage.PrefixBlobRecord}
+	itemsPerBatch := s.batchDeleteItemLimit()
+	var batch []*storage.DeleteInfo
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if err := s.setPrunedHeight(height); err != nil {
+		return err
+	}
+
+	if err := s.db.View(func(txn *badger.Txn) error {
+		it := txn.NewIterator(badger.IteratorOptions{
+			PrefetchValues: false,
+			Prefix:         blobRecordPrefix,
+		})
+		defer it.Close()
+
+		// iterate over blob records, calling pruneCallback for any CIDs that should be pruned
+		// and cleaning up the corresponding tracker records
+		for it.Seek(blobRecordPrefix); it.ValidForPrefix(blobRecordPrefix); it.Next() {
+			blobRecordItem := it.Item()
+			blobRecordKey := blobRecordItem.Key()
+
+			blockHeight, blobCid, err := storage.ParseBlobRecordKey(blobRecordKey)
+			if err != nil {
+				return fmt.Errorf("malformed blob record key %v: %w", blobRecordKey, err)
+			}
+
+			// iteration occurs in key order, so block heights are guaranteed to be ascending
+			if blockHeight > height {
+				break
+			}
+
+			dInfo := &storage.DeleteInfo{
+				Cid:    blobCid,
+				Height: blockHeight,
+			}
+
+			var latestHeight uint64
+			err = operation.RetrieveTrackerLatestHeight(blobCid, &latestHeight)(txn)
+			if err != nil {
+				return fmt.Errorf("failed to get latest height entry for Cid %s: %w", blobCid.String(), err)
+			}
+
+			// a blob is only removable if it is not referenced by any blob tree at a higher height
+			if latestHeight < blockHeight {
+				// this should never happen
+				return fmt.Errorf(
+					"inconsistency detected: latest height recorded for Cid %s is %d, but blob record exists at height %d",
+					blobCid.String(), latestHeight, blockHeight,
+				)
+			}
+
+			// the current block height is the last to reference this CID, prune the CID and remove
+			// all tracker records
+			if latestHeight == blockHeight {
+				if err := s.pruneCallback(blobCid); err != nil {
+					return err
+				}
+				dInfo.DeleteLatestHeightRecord = true
+			}
+
+			// remove tracker records for pruned heights
+			batch = append(batch, dInfo)
+			if len(batch) == itemsPerBatch {
+				if err := s.batchDelete(batch); err != nil {
+					return err
+				}
+				batch = nil
+			}
+		}
+
+		if len(batch) > 0 {
+			if err := s.batchDelete(batch); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// this is a good time to do garbage collection
+	if err := s.db.RunValueLogGC(0.5); err != nil {
```

Comment:
make the `0.5` a constant. I think we should go with `0.25` as an initial default

Reply:
[2b609d4](https://github.com/onflow/flow-go/pull/6277/commits/2b609d4dabf9025204834b95321ba15fe8859f86)